### PR TITLE
added support for idempotency key

### PR DESCRIPTION
**Pull request by mschielmann**

### CODE REVIEW

File: spec/integration/idempotencyKeyChargesSpec.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        const requestOptions = new RequestOptions();
+        requestOptions.setIdempotencyKey(crypto.randomUUID())
+        const chargeReq = charges.charge({card: cards.card()})
+
+        // when
+        const createdCharge = await api.charges.create(chargeReq, requestOptions)
+        const sameCharge = await api.charges.create(chargeReq, requestOptions)
+
+        // then
+        expect(sameCharge.id).toEqual(createdCharge.id)
```
Comment (rave-shift4) [inline spec/integration/idempotencyKeyChargesSpec.js]: Can we check here attendance of header: `idempotent-replayed`?

File: spec/integration/idempotencyKeyChargesSpec.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+    it('should create one charge without duplicates on multiple requests if idempotency key is used', async () => {
+        // given
+        const requestOptions = new RequestOptions();
+        requestOptions.setIdempotencyKey(crypto.randomUUID())
+        const chargeReq = charges.charge({card: cards.card()})
+
+        // when
+        const createdCharge = await api.charges.create(chargeReq, requestOptions)
+        const sameCharge = await api.charges.create(chargeReq, requestOptions)
```
Comment (rave-shift4) [inline spec/integration/idempotencyKeyChargesSpec.js]: And like in prev: consider the tests for each modified method ;P

File: lib/communicator.js [hunk 3]
```diff
@@ -16,8 +16,12 @@ module.exports = class Communicator {
     this.url = removeTrailingSlash(url)
   }
 
-  async post (path, body) {
-    return this.call('POST', path, JSON.stringify(body), { 'Content-Type': 'application/json' })
+  async post (path, body, requestOptions) {
+    const headers = { 'Content-Type': 'application/json' };
+    if (!!requestOptions && requestOptions.hasIdempotencyKey()) {
+      headers['Idempotency-Key'] = requestOptions.getIdempotencyKey()
```
Comment (el-dot) [inline lib/communicator.js]: it should be simple JS object, passed without constructor. 

JS developers are not used to use verbose class, constructors and DTOs for passing simple parameters.

**Review by el-dot — CHANGES_REQUESTED** [review]

File: lib/communicator.js (hunk 3 above)
Comment (mschielmann) [reply lib/communicator.js]: done.

File: spec/integration/idempotencyKeyChargesSpec.js (hunk 1 above)
Comment (el-dot) [reply spec/integration/idempotencyKeyChargesSpec.js]: We don't pass headers in this SDK to the SDK user, closing for now

**Pull request merged**